core/crdt/pncounter.go

```diff
@@ -0,0 +1,213 @@
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package crdt
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/fxamacker/cbor/v2"
+	dag "github.com/ipfs/boxo/ipld/merkledag"
+	ds "github.com/ipfs/go-datastore"
+	ipld "github.com/ipfs/go-ipld-format"
+	"github.com/ugorji/go/codec"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/core"
+	"github.com/sourcenetwork/defradb/datastore"
+	"github.com/sourcenetwork/defradb/db/base"
+	"github.com/sourcenetwork/defradb/errors"
+)
+
+var (
+	// ensure types implements core interfaces
+	_ core.ReplicatedData = (*PNCounterRegister)(nil)
+	_ core.Delta          = (*PNCounterDelta)(nil)
+)
+
+// PNCounterDelta is a single delta operation for an PNCounterRegister
+type PNCounterDelta struct {
+	DocKey    []byte
+	FieldName string
+	Priority  uint64
+	// SchemaVersionID is the schema version datastore key at the time of commit.
+	//
+	// It can be used to identify the collection datastructure state at the time of commit.
+	SchemaVersionID string
+	Data            []byte
+	FieldValue      client.FieldValue `json:"-"` // should not be marshalled
+}
+
+// GetPriority gets the current priority for this delta.
+func (delta *PNCounterDelta) GetPriority() uint64 {
+	return delta.Priority
+}
+
+// SetPriority will set the priority for this delta.
+func (delta *PNCounterDelta) SetPriority(prio uint64) {
+	delta.Priority = prio
+}
+
+// Marshal encodes the delta using CBOR.
+func (delta *PNCounterDelta) Marshal() ([]byte, error) {
+	b, err := delta.FieldValue.Bytes()
+	if err != nil {
+		return nil, err
+	}
+	delta.Data = b
+
+	h := &codec.CborHandle{}
+	buf := bytes.NewBuffer(nil)
+	enc := codec.NewEncoder(buf, h)
+	err = enc.Encode(delta)
+	if err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// PNCounterRegister, Last-Writer-Wins Register, is a simple CRDT type that allows set/get
+// of an arbitrary data type that ensures convergence.
+type PNCounterRegister struct {
+	baseCRDT
+}
+
+// NewPNCounterRegister returns a new instance of the PNCounter with the given ID.
+func NewPNCounterRegister(
+	store datastore.DSReaderWriter,
+	schemaVersionKey core.CollectionSchemaVersionKey,
+	key core.DataStoreKey,
+	fieldName string,
+) PNCounterRegister {
+	return PNCounterRegister{newBaseCRDT(store, key, schemaVersionKey, fieldName)}
+}
+
+// Value gets the current register value
+func (reg PNCounterRegister) Value(ctx context.Context) ([]byte, error) {
+	valueK := reg.key.WithValueFlag()
+	buf, err := reg.store.Get(ctx, valueK.ToDS())
+	if err != nil {
+		return nil, err
+	}
+	return buf, nil
+}
+
+// Set generates a new delta with the supplied value
+func (reg PNCounterRegister) Set(value client.FieldValue) *PNCounterDelta {
```

Comment:
`Set` actually isn't part of the `ReplicatedData` interface. No delta-mutators are defined on the interface, because they are all CRDT specific.

I would actually prefer this gets resolved in this PR, as this will be setting a standard for future CRDT types, and we should make it clear that the mutators do need to be CRDT specific, rather than following the one existing structure fot the `LWWRegister` type.

Reply:
We can probably rename it to `NewDelta`

Reply:
Why would it be called `NewDelta`? Its a delta-mutator, it should be named what mutation it does. `Set` is very intentionally called `Set` befacuse the mutation operation on a register is "setting" it. Counters can "increment" or "decrement".

Reply:
That function doesn't "do" anything though. It only instantiates a delta type. The "doing" happens in the merge function.

Logically, the increment/decrement, set or other mutation call makes sense on the MerkleCRDT type. Within that function though, the call to `Set` that returns a delta type is confusing. `NewDelta` makes a lot more sense.

Reply:
I understand what you're saying. It is still confusing to read. If we don't mind that confusion and want to keep that nomenclature, I'll make them match.

Reply:
The confusion that I'm referring to here is not that I personally am confused when reading it. I'm mentioning this as if I was in the shoes of someone trying to understand what is happening but doesn't understand everything about CRDTs. The naming that you prefer adds confusion because it isn't clearly relatable to what is happening in the function.

Reply:
Noone should be touching this code without understanding the CRDTs. There's even an entire separate `README` for this CRDT package specifically because they are unusual data structures that require nuance.

Reply:
Sounds good 👍

Reply:
✅

Reply:
I do not agree that the only way for CRDTs to apply a new value is through a crdt-type-specific methods (like `Set`, `Increment`...). Yes, we can define them on the corresponding structs, but why do we want to force them? 

Why should `db.Collection` care about specific way we store values and different CRDT types? It should get a new value (`client.FieldValue`, `[]byte` or `any`) and passes it further to whatever CRDT it deals with.

> The Mutator funcs are named after the mutation operation they are implementing.

@jsimnz  Even is there is such a convention we don't have to stick to it. I'm sure our code will benefit from a polymorphic approach. Otherwise will have ugly `switch` cases throughout the code.

I would propose a name `ApplyValue` for the interface method and every CRDT will decide how to apply a value.

Reply:
> I do not agree that the only way for CRDTs to apply a new value is through a crdt-type-specific methods (like Set, Increment...). Yes, we can define them on the corresponding structs, but why do we want to force them?
>
> Why should db.Collection care about specific way we store values and different CRDT types? It should get a new value (client.FieldValue, []byte or any) and passes it further to whatever CRDT it deals with.

Just noting that I fully agree with what Islam wrote here.  And I agree with Fred wrote earlier RE confusion.  

Yes much of the literature RE CRDTs is very academic, but that doesn't mean that our code needs to be unreadable to anyone but academics.  The core concepts aren't that complicated or even new to programmers really.